### PR TITLE
refactor(lightning): refractored pytorch code to pytorch-lightning

### LINKED ISSUE #1: Optimize pytorch run as good as possible.
- [x] Asynchronous data loading and augmentation enabled.
- [x] cuDNN autotuner enabled.
- [x] Automatic mixed precision, i.e use of float16 in tensorcores.
- [ ] Increase batch-size due to reduced memory usage.
    - [ ] Tune learning rate.
    - [ ] Add learning rate warmup.
    - [x] Add learning decay.
    - [ ] Add weight decay.
    - [ ] Use optimizer for large-barch training(LARS, LAMB, NVLAMB, NovoGrad).
- [x] Disable bias for convolutions directly followed by a batch norm.
- [x] Use parameter.grad = None.
- [x] Disbale debut APIs in production-mode.
- [ ] Opt for [DistributedDataParallel](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html)
instead of [DataParallel](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html) to improve Multi-GPU efficiency.
    - [ ] Load-balance workload on multiple GPUs.
- [ ] Use fused building blocs from [APEX](https://github.com/NVIDIA/apex).
- [ ] Checkpoint to recompute intermediates.
- [ ] Fuse pointwise operations.

### LINKED ISSUE #4: Implement AUC - ROC as a metric?
https://towardsdatascience.com/understanding-auc-roc-curve-68b2303cc9c5

### LINKED ISSUE #7: Adopting Lightning in your code
https://towardsdatascience.com/7-tips-for-squeezing-maximum-performance-from-pytorch-ca4a40951259

### LINKED ISSUE #18: train_net is too complex (16) (MC0001)
#### Occurred on:
+ **Commit**:  2f76376d8dfa0dca4da6fe982ca29f35e57a4c12
+ **File**:    [train.py](https://github.com/gil-uav/semantic-image-segmentation/blob/2f76376d8dfa0dca4da6fe982ca29f35e57a4c12/train.py)
+ **LineNum**: [28](https://github.com/gil-uav/semantic-image-segmentation/blob/2f76376d8dfa0dca4da6fe982ca29f35e57a4c12/train.py#L28)
+ **Code**:    `def train_net(`

    
#### Currently on:
+ **Commit**:  8dc9706a7e77cfc1ae940b78a787b56feb13ae45
+ **File**:    [train.py](https://github.com/gil-uav/semantic-image-segmentation/blob/8dc9706a7e77cfc1ae940b78a787b56feb13ae45/train.py)
+ **LineNum**: [28](https://github.com/gil-uav/semantic-image-segmentation/blob/8dc9706a7e77cfc1ae940b78a787b56feb13ae45/train.py#L28)

### LINKED ISSUE #20: Improve image visualization to tensorboard.

### LINKED ISSUE #21: Optimize with these tips.
https://towardsdatascience.com/7-tips-for-squeezing-maximum-performance-from-pytorch-ca4a40951259

### PR DESCRIPTION
**Pull request by vegovs**
Complete refractor in order to make research more streamless.

BREAKING CHANGE: Complete refractor.

### CODE REVIEW

File: unet/unet_model.py [hunk 1]
```diff
@@ -64,3 +91,208 @@ def forward(self, x):
...
+            drop_last=True,
+        )
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(
+            self.parameters(), lr=(self.lr or self.learning_rate)
+        )
+        return [optimizer]
+
+    def loss_funciton(self, input, target):
```
Comment (vegovs) [inline unet/unet_model.py:143]: Codacy found an issue: [Redefining built-in 'input'](https://app.codacy.com/gh/gil-uav/semantic-image-segmentation/pullRequest?prid=6243237)

**Comment by vegovs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- unet/unet_model.py  4
         

Complexity decreasing per file
==============================
+ train.py  -5
         

Clones added
============
- unet/unet_model.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gil-uav/semantic-image-segmentation/pullRequest?prid=6243237&bid=20336541)

**Comment by vegovs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- unet/unet_model.py  4
         

Complexity decreasing per file
==============================
+ train.py  -5
         

Clones added
============
- unet/unet_model.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gil-uav/semantic-image-segmentation/pullRequest?prid=6243237&bid=20336541)
